### PR TITLE
Added equipment and transfer

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -59,6 +62,7 @@ pub struct PtObjects {
     datasets_to_vehicle_journeys: OneToMany<Dataset, VehicleJourney>,
     companies_to_vehicle_journeys: OneToMany<Company, VehicleJourney>,
     vehicle_journeys_to_stop_points: ManyToMany<VehicleJourney, StopPoint>,
+    pub stop_points_to_stop_points: ManyToMany<StopPoint, StopPoint>,
```
Comment (TeXitoi) [inline src/lib.rs]:
```rust
transfers_to_stop_points: ManyToMany<Transfer, StopPoint>,
```
(no pub)

File: src/collection.rs [hunk 2]
```diff
@@ -36,11 +36,30 @@ impl<T> PartialOrd for Idx<T> {
...
+        Collection::new(Vec::default())
+    }
+}
+
+impl<T> ops::Deref for Collection<T> {
+    type Target = Vec<T>;
+    fn deref(&self) -> &Vec<T> {
+        &self.objects
+    }
+}
```
Comment (TeXitoi) [inline src/collection.rs]: not this one, you want
```rust
impl<T& ops::Deref for CollectionWithId<T> {
    type Target = Collection<T>;
    fn deref(&self) -> &Target {
        &self.collection
    }
}
```
```

File: src/collection.rs [hunk 3]
```diff
@@ -146,22 +195,22 @@ impl<T> Collection<T> {
...
+        self.collection.objects.len()
     }
 }
 
-impl<T> ops::Index<Idx<T>> for Collection<T> {
+impl<T> ops::Index<Idx<T>> for CollectionWithId<T> {
     type Output = T;
     fn index(&self, index: Idx<T>) -> &Self::Output {
-        &self.objects[index.get()]
+        &self.collection.objects[index.get()]
```
Comment (TeXitoi) [inline src/collection.rs]: With deref, you should not need these `.collection`

File: src/lib.rs [hunk 4]
```diff
@@ -81,6 +85,14 @@ impl PtObjects {
                 )
             })
             .collect();
+        let mut forward_sp_to_sp = BTreeMap::default();
+        c.transfers.iter().for_each(|obj| {
+            forward_sp_to_sp
+                .entry(c.stop_points.get_idx(&obj.from_stop_id).unwrap())
+                .or_insert_with(IdxSet::default)
+                .insert(c.stop_points.get_idx(&obj.to_stop_id).unwrap());
+        });
```
Comment (TeXitoi) [inline src/lib.rs]: To change as that's not the functionality we want.

File: src/objects.rs [hunk 5]
```diff
@@ -468,6 +468,47 @@ impl Id<Comment> for Comment {
     }
 }
 
+#[derive(Derivative)]
+#[derivative(Default(bound = ""))]
```
Comment (TeXitoi) [inline src/objects.rs]:
```rust
#[derivative(Default)]
```

File: src/objects.rs [hunk 6]
```diff
@@ -468,6 +468,47 @@ impl Id<Comment> for Comment {
...
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Disponibility {
+    #[derivative(Default)]
+    #[serde(rename = "0")]
```
Comment (TeXitoi) [inline src/objects.rs]:
```rust
#[derivative(Default), serde(rename = "0")]
```

File: src/objects.rs [hunk 7]
```diff
@@ -468,6 +468,47 @@ impl Id<Comment> for Comment {
...
+    InformationNotDisponible,
+    #[serde(rename = "1")] Disponible,
+    #[serde(rename = "2")] NotDisponible,
```
Comment (TeXitoi) [inline src/objects.rs]: I don't like these names, they sound very franglais.

File: src/objects.rs (hunk 6 above)
Comment (datanel) [reply src/objects.rs]: this does not work
```
error: proc-macro derive panicked
   --> src/objects.rs:473:21
    |
473 | #[derive(Serialize, Deserialize, Debug)]
    |                     ^^^^^^^^^^^
    |
    = help: message: called `Result::unwrap()` on an `Err` value: "failed to parse derive input: \"#[derivative(Default)]\\npub enum Disponibility {\\n\\n  #[derivative\\n    ( Default ) , serde ( rename = \\\"0\\\" )]\\n    InformationNotDisponible,\\n\\n    #[serde(rename = \\\"1\\\")]\\n    Disponible,\\n\\n    #[serde(rename = \\\"2\\\")]\\n    NotDisponible,\\n}\""
```

Comment (TeXitoi) [reply src/objects.rs]: My bad it's not allowed.

File: src/collection.rs (hunk 2 above)
Comment (TeXitoi) [reply src/collection.rs]: Please remove this deref.

File: src/collection.rs [hunk 8]
```diff
@@ -91,12 +146,12 @@ impl<T: Id<T>> Collection<T> {
...
 pub struct RefMut<'a, T: 'a + Id<T>> {
     idx: Idx<T>,
-    collection: &'a mut Collection<T>,
+    collection: &'a mut CollectionWithId<T>,
     old_id: String,
 }
 impl<'a, T: Id<T>> ops::DerefMut for RefMut<'a, T> {
     fn deref_mut(&mut self) -> &mut T {
-        &mut self.collection.objects[self.idx.get()]
+        &mut self.collection.collection.objects[self.idx.get()]
```
Comment (TeXitoi) [inline src/collection.rs]: you can remove one collection here

File: src/collection.rs [hunk 9]
```diff
@@ -146,22 +202,31 @@ impl<T> Collection<T> {
...
+        ::std::mem::replace(&mut self.collection.objects, Vec::new())
     }
 
     pub fn len(&self) -> usize {
         self.objects.len()
     }
 }
 
-impl<T> ops::Index<Idx<T>> for Collection<T> {
+impl<T> ops::Index<Idx<T>> for CollectionWithId<T> {
```
Comment (TeXitoi) [inline src/collection.rs]: This one should be on Collection

File: src/objects.rs [hunk 10]
```diff
@@ -468,6 +468,47 @@ impl Id<Comment> for Comment {
     }
 }
 
+#[derive(Derivative)]
+#[derivative(Default)]
+#[derive(Serialize, Deserialize, Debug)]
```
Comment (TeXitoi) [inline src/objects.rs]:
```rust
#[derive(Serialize, Deserialize, Debug, Derivative)]
#[derivative(Default)]
```

File: src/collection.rs (hunk 8 above)
Comment (datanel) [reply src/collection.rs]: if I remove
```
&mut self.collection.objects[self.idx.get()]
    |              ^^^^^^^^^^^^^^^^^^^^^^^ cannot mutably borrow immutable field
```

Comment (TeXitoi) [reply src/collection.rs]: Ha oui, OTAN pour moi.

File: src/collection.rs [hunk 11]
```diff
@@ -120,18 +168,19 @@ impl<'a, T: Id<T>> Drop for RefMut<'a, T> {
...
         }
     }
 }
 
-impl<T> Collection<T> {
+impl<T> CollectionWithId<T> {
     pub fn iter(&self) -> Iter<T> {
-        self.objects
+        self.collection
+            .objects
```
Comment (TeXitoi) [inline src/collection.rs]: You don't need this one as it is implemented on Collection<T>

**Pull request merged**